### DISCUSSION TITLE
NO X-AXIS MOVEMENT ON NANOELS H4

### DISCUSSION THREAD

**Discussion by RGUERI1** [kachurovskiy/nanoels#126] [Q&A] [answered]
Hi, I just got my ELS 4 hooked up and all components mounted. Made all ratio mods in the code and uploaded. The Z-axis works as intended but I don't have any movement out of my X-axis. I tried swapping the plugs on the back of the NANO and my X-axis works, albeit with the LEFT/RIGHT arrows. With the plugs swapped, the UP/Down arrows don't move the Z-AXIS either. 
That tells me that the problem has to be on the SOFTWARE side or within the controller itself? 
Am I missing something? Any thoughts/comments are appreciated. 

These are my current settings:
```
// Define your hardware parameters here. Don't remove the ".0" at the end.
#define ENCODER_STEPS 1200 // 600 step spindle optical rotary encoder
#define ENCODER_BACKLASH 3 // Numer of impulses encoder can issue without movement of the spindle

// Spindle rotary encoder pins. Swap values if the rotation direction is wrong.
#define ENC_A 7
#define ENC_B 15

// Main lead screw (Z) parameters.
#define MOTOR_STEPS_Z 200.0
#define SCREW_Z_DU 21170.0 // 12TPI=2.117MM PITCH lead screw in deci-microns (10^-7 of a meter)
#define SPEED_START_Z (2 * MOTOR_STEPS_Z) // Initial speed of a motor, steps / second.
#define ACCELERATION_Z (40 * MOTOR_STEPS_Z) // Acceleration of a motor, steps / second ^ 2.
#define SPEED_MANUAL_MOVE_Z (6 * MOTOR_STEPS_Z) // Maximum speed of a motor during manual move, steps / second.
#define INVERT_Z false // change (true/false) if the carriage moves e.g. "left" when you press "right".
#define NEEDS_REST_Z false // Set to false for closed-loop drivers, true for open-loop.
#define MAX_TRAVEL_MM_Z 300 // Lathe bed doesn't allow to travel more than this in one go, 30cm / ~1 foot
#define BACKLASH_DU_Z 6500 // 0.65mm backlash in deci-microns (10^-7 of a meter)
#define NAME_Z 'Z' // Text shown on screen before axis position value, GCode axis name

// Cross-slide lead screw (X) parameters.
#define MOTOR_STEPS_X 600 // 200 steps at 3x reduction
#define SCREW_X_DU 10000.0 // 1.00mm lead screw with 3x reduction in deci-microns (10^-7) of a meter
#define SPEED_START_X (MOTOR_STEPS_X) // Initial speed of a motor, steps / second.
#define ACCELERATION_X (10 * MOTOR_STEPS_X) // Acceleration of a motor, steps / second ^ 2.
#define SPEED_MANUAL_MOVE_X (2 * MOTOR_STEPS_X) // Maximum speed of a motor during manual move, steps / second.
#define INVERT_X true // change (true/false) if the carriage moves e.g. "left" when you press "right".
#define NEEDS_REST_X false // Set to false for all kinds of drivers or X will be unlocked when not moving.
#define MAX_TRAVEL_MM_X 100 // Cross slide doesn't allow to travel more than this in one go, 10cm
#define BACKLASH_DU_X 1500 // 0.15mm backlash in deci-microns (10^-7 of a meter)
#define NAME_X 'X' // Text shown on screen before axis position value, GCode axis name

// Manual stepping with left/right/up/down buttons. Only used when step isn't default continuous (1mm or 0.1").
#define STEP_TIME_MS 500 // Time in milliseconds it should take to make 1 manual step.
#define DELAY_BETWEEN_STEPS_MS 80 // Time in milliseconds to wait between steps.
```

**Comment by kachurovskiy** [comment]
Hi @RGUERI1, comparing your code with the one in the repo:

![image](https://github.com/kachurovskiy/nanoels/assets/517919/c4a819e4-596a-4564-8b97-a9daddf7304a)

Consider adding `.0` at the end in the line 2 and 22, this could avoid some problems down the road. I will stop using `#define` in the next versions and it won't be a problem in the future but for now it's better to include the floating point 0.

The fact that swapping the X and Z plugs on the back makes X move tells that motor drivers are OK and wired well. Since you're using `NEEDS_REST_... false` it's not about the enable high/low either.

It would be interesting to know what voltages do you get on the X pins.

Each controller that I ship gets tested - all buttons clicked, motion of both axes both directions, encoder back/forth, etc. - but there's always something that can go wrong.

As far as an immediate mitigation goes, try swapping `A1` terminal with `X` in the code. To do this, replace [lines 99-109](https://github.com/kachurovskiy/nanoels/blob/main/h4/h4.ino#L99C16-L109) with the code below:

```
#define X_ENA 9
#define X_DIR 10
#define X_STEP 11

#define BUZZ 4
#define SCL 5
#define SDA 6

#define A11 8
#define A12 19
#define A13 20
```

Please let me know what you think. Thanks!

**Reply by kachurovskiy** [reply]
Another option is that I can mail you a replacement main board with a screen - but it will take a month or so until it arrives.

**Comment by RGUERI1** [comment]
Maxim,
I ran out of time yesterday, hopefully get back on it this evening. I'll
keep you posted as to what I find. If I have to use the aux port, it's no
big deal.
Thanks

**Reply by RGUERI1** [reply, marked as answer]
I figured it was something stupid. As I was going in to paste the code you sent above, I noticed the line #define X_ENA 8 and realized it should have been #define X_ENA 18. Also explains why I had no voltage on that pin! I must have erased the 1 at some point.
Problem solved, thanks for the help.
